nucleus/common/common-util/src/main/java/com/sun/enterprise/util/io/FileUtils.java

```diff
@@ -561,43 +548,54 @@ public static void deleteFileNowOrLater(File f) {
      * Java Runtime. Non-empty directories will be deleted recursively.
      * @param f the file to delete on JVM shutdown
      */
-    public static void deleteOnExitRecursively(File f) {
+    public static void deleteOnExit(File f) {
         if (f != null && f.exists()) {
             final SecurityManager securityManager = System.getSecurityManager();
             if (securityManager != null) {
                 securityManager.checkDelete(f.getAbsolutePath());
             }
-            FILES_TO_DELETE_ON_SHUTDOWN.add(f);
+            FILE_DELETION_ON_EXIT.add(f);
         }
     }
 
-    private static final Set<File> FILES_TO_DELETE_ON_SHUTDOWN = new LinkedHashSet<>();
+    private static final FileDeletionThread FILE_DELETION_ON_EXIT = new FileDeletionThread();
 
     static {
-        Runtime.getRuntime().addShutdownHook(
-            new Thread() {
-                @Override
-                public void run() {
-                    for (File f : FILES_TO_DELETE_ON_SHUTDOWN) {
-                        delete(f);
-                    }
-                    FILES_TO_DELETE_ON_SHUTDOWN.clear();
-                }
-                private void delete(File file) {
-                    if (file.exists()) {
-                        if (file.isDirectory()) {
-                            final File[] files = file.listFiles();
-                            if (files != null) {
-                                for (File f : files) {
-                                    delete(f);
-                                }
-                            }
-                        }
-                        file.delete();
-                    }
+        Runtime.getRuntime().addShutdownHook(FILE_DELETION_ON_EXIT);
+    }
+
+    private static final class FileDeletionThread extends Thread {
+
+        private static final Logger LOG = Logger.getLogger(FileDeletionThread.class.getName());
+        private final Set<String> filesToDelete = new LinkedHashSet<>();
```

Comment:
Is it sure only one thread at a time will access field filesToDelete? If not, there is a race condition that may cause ConcurrentModificationException during clean up (line 580).

Reply:
Good point. I will address this.

Reply:
I have addressed it.